server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,253 @@
+That `update_id` can be provided to re-attach to an in-flight or completed
+Update and obtain the associated UpdateResult in the form of an
+`UpdateWorkflowResponse`.
+
+
+```protobuf
+service WorkflowService {
+    rpc UpdateWorkflow(UpdateWorkflowRequest) returns (UpdateWorkflowResponse){}
+    rpc PollUpdateWorkflowResponse(PollUpdateWorkflowResponseRequest) returns (UpdateWorkflowResponse){}
+}
+
+message UpdateWorkflowRequest {
+    temporal.api.common.v1.Header header = 1;
+    string request_id = 2;
+    string namespace = 3;
+    temporal.api.common.v1.WorkflowExecution execution = 4;
+    string update_name = 5;
+    temporal.api.common.v1.Payloads input = 6;
+    UpdateWorkflowResultAccessStyle result_access_style = 7;
+}
+
+message UpdateWorkflowResponse {
+    bytes update_id = 1;
+    UpdateWorkflowResultType result_type = 2;
+    UpdateWorkflowSuccess success = 3;
+    UpdateWorkflowFailure failure = 4;
+}
+
+enum UpdateWorkflowResultAccessStyle {
+    UPDATE_WORKFLOW_RESULT_ACCESS_STYLE_UNSPECIFIED = 0;
+    UPDATE_WORKFLOW_RESULT_ACCESS_STYLE_INLINE = 1;
+    UPDATE_WORKFLOW_RESULT_ACCESS_STYLE_POLL = 2;
+}
+
+enum UpdateWorkflowResultType {
+    UPDATE_WORKFLOW_RESULT_TYPE_UNSPECIFIED = 0;
+    UPDATE_WORKFLOW_RESULT_TYPE_IN_PROGRESS = 1;
+    UPDATE_WORKFLOW_RESULT_TYPE_SUCCESS = 2;
+    UPDATE_WORKFLOW_RESULT_TYPE_FAILURE = 3;
+}
+
+message PollUpdateResponseRequest {
+    bytes update_id = 1;
+}
+
+message UpdateWorkflowSuccess {
+  temporal.api.common.v1.Payloads output = 1;
+}
+
+message UpdateWorkflowFailure {
+  temporal.api.common.v1.Failure output = 1;
+}
+```
+
+## Revisiting Use Cases With Propsed API
+
+### Short Update
+
+Client calls `UpdateWorkflow` with `UpdateWorkflowRequest.result_access_style`
+set to `UPDATE_WORKFLOW_RESULT_ACCESS_STYLE_INLINE`. The UpdateResponse is returned
+as part of the gRPC response.
+
+### Uncertain Duration Update
+
+Two options:
+
+1. Client calls `UpdateWorkflow` with `UpdateWorkflowRequest.result_access_style`
+   set to `UPDATE_WORKFLOW_RESULT_ACCESS_STYLE_INLINE`. If the UpdateResponse is
+   slow, client can retry safely with the same `request_id`.
+2. Client follows the steps for the "Known Long Update" use case.
+
+### Known Long Update
+
+Client calls `UpdateWorkflow` with `UpdateWorkflowRequest.result_access_style`
+set to `UPDATE_WORKFLOW_RESULT_ACCESS_STYLE_POLL`. An UpdateID is returned as
+part of the gRPC respons and if that same gRPC response indicates that the
+Update is still in progress (i.e. `UpdateWorkflowResponse.result_type ==
+UPDATE_WORKFLOW_RESPONSE_TYPE_IN_PROGRESS`), the client can then use
+`PollUpdateWorkflowResponse` to wait for Update completion.
+
+### Invalid Update
+
+If the UpdateRequest is rejected by the workflow execution while the RPC client
+is still connected, a suitable gRPC error can be returned synchronously. Note
+however rejected UpdateRequests are _not_ recorded in workflow history.
```

Comment:
Arguably they should be recorded in mutable state though in order to deliver the response asynchronously.